### ISSUE TITLE
CHTable at the JITaaS server should not be global

### ISSUE THREAD

**Issue by mpirvu** [eclipse/openj9#3724]
Currently, the JITaaS server has only one object of type `TR_JITaaSServerPersistentCHTable`
but this object knows how to get to the real per-client chtables. One problem I see is that the same lock is obtained for all the clients, which creates more contention than needed when multiple clients are served in parallel by the same JITaaS. We should probably use several objects that are embedded in the client session and each onbject should have their own lock.
Moreover, obtaining the CHTable mutex is done with the same complicated routine used for non-JITaaS (which has to deal with class loading). In the JITaaS case we just need a simple lock.

**Comment by mpirvu** [comment]
@harryyu1994 Harry, as discussed, please start working on this issue. Thanks

**Comment by harryyu1994** [comment]
I think the reason Noah implemented a global `TR_JITaaSServerPersistentCHTable` object is to have minimum changes to non-JITaaS parts. `TR_JITaaSServerPersistentCHTable` extends `TR_PersistentCHTable` and is meant to be queried the same way in non-JITaaS codegen and JITaaSServer codegen. (Something like this is shared by non-JITaaS & JITaaSServer: `self()->comp()->getPersistentInfo()->getPersistentCHTable()->findClassInfoAfterLocking(classPointer, self()->comp());`)

So in my opinion, giving each client their own chtable lock and using the per-client lock along with the per-client chtable is sufficient. 

Hi Marius, please let me know what you think @mpirvu .

**Comment by mpirvu** [comment]
If we always go through `getPersistentCHTable()` can we change the definition of that method to return the 'per client' CHTable?

**Comment by harryyu1994** [comment]
I think that could work, I will come up with an implementation

**Comment by harryyu1994** [comment]
Here is the definition of `getPersistentCHTable()` in `compiler/env/J9PersistentInfo.hpp`, I failed to see how we can access the `clientSessionData` from here.
```
TR_PersistentCHTable * getPersistentCHTable() { return _persistentCHTable; }
void setPersistentCHTable(TR_PersistentCHTable *table) { _persistentCHTable = table; }
```

**Comment by mpirvu** [comment]
You could use `TR::compInfoPT->getClientData()` but you need another test to make sure that this happens only for JITaaS server. Could we extend PersistentMemory to create a specialized version just for JITaaS?

**Comment by harryyu1994** [comment]
Working on this issue now.

**Comment by harryyu1994** [comment]
Going to look into extending PersistentMemory for JITaaS. It may be a good idea to extend as I'm seeing a few JITaaS-specific things in the existing PersistentMemory, we might as well just move them altogether.

**Comment by harryyu1994** [comment]
With the way the class `TR_PersistentMemory` is defined (in omr), I don't think we have a clean solution to extend `PersistentMemory` & `PersistentInfo`.
```
class TR_PersistentMemory : public TR_MemoryBase
   {
   ...
   TR::PersistentInfo * getPersistentInfo() { return &_persistentInfo; }
   TR::PersistentInfo _persistentInfo;
    ...
    }
```
- the goal of extending TR_PersistentMemory is so that we can have a specialized version of `TR::PersistentInfo` (`TR::JITaaSPersistentInfo`) that overrides `getPersistentCHTable()` to give us the correct CHTable.
- However, instead of a pointer to TR::PersistentInfo as a member field, we have a whole TR::PersistentInfo class inside of TR_PersistentMemory. Which means even if we extend TR_PersistentMemory, we can't change this field.
- So that left us with the solution of changing the existing `getPersistentCHTable()` in `PersistentInfo` to give us the right CHTable.

**Comment by mpirvu** [comment]
FYI: @dmitry-ten

**Comment by mpirvu** [comment]
Implemented in #10508
